sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -699,8 +732,12 @@ case class Substring(str: Expression, pos: Expression, len: Expression)
       if (posEval != null) {
         val lenEval = len.eval(input)
         if (lenEval != null) {
-          stringEval.asInstanceOf[UTF8String]
-            .substringSQL(posEval.asInstanceOf[Int], lenEval.asInstanceOf[Int])
+          str.dataType match {
+            case StringType => stringEval.asInstanceOf[UTF8String]
+              .substringSQL(posEval.asInstanceOf[Int], lenEval.asInstanceOf[Int])
+            case BinaryType => Substring.subStringBinarySQL(stringEval.asInstanceOf[Array[Byte]],
```

Comment:
`Substring#subStringBinarySQL` is created on purpose for the index in byte, but the `substringSQL` is operate on the index of code point.